tests/loramac/main.c

```diff
@@ -0,0 +1,333 @@
+( (____  _____ ____ _| |_ _____  ____| |__
+ \____ \| ___ |    (_   _) ___ |/ ___)  _ \
+ _____) ) ____| | | || |_| ____( (___| | | |
+(______/|_____)_|_|_| \__)_____)\____)_| |_|
+    (C)2013 Semtech
+Description: LoRaMac classA device implementation
+License: Revised BSD License, see LICENSE.TXT file include in the project
+Maintainer: Miguel Luis and Gregory Cristian
+*/
+#include <errno.h>
+#include <stdio.h>
+#include <math.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "shell.h"
+#include "shell_commands.h"
+#include "thread.h"
+#include "xtimer.h"
+#include "periph/rtc.h"
+#include "random.h"
+
+#include "common.h"
+#include "board.h"
+
+#include "net/gnrc/netdev2.h"
+#include "net/netdev2.h"
+#include "sx1276_params.h"
+#include "sx1276_netdev.h"
+#include "net/gnrc/lorawan/lorawan.h"
+#include "net/lorawan/hdr.h"
+#include "net/gnrc/netreg.h"
+
+#define GNRC_LORA_MSG_QUEUE 16
+
+/*!
+ * LoRaWAN confirmed messages
+ */
+#ifndef LORAWAN_CONFIRMED_MSG_ON 
+    #define LORAWAN_CONFIRMED_MSG_ON                    false
+#endif
+
+/*!
+ * LoRaWAN Adaptive Data Rate
+ *
+ * \remark Please note that when ADR is enabled the end-device should be static
+ */
+#ifndef LORAWAN_ADR_ON 
+    #define LORAWAN_ADR_ON                              0
+#endif
+
+/*!
+ * LoRaWAN application port
+ */
+#ifndef LORAWAN_APP_PORT 
+    #define LORAWAN_APP_PORT                            2
+#endif
+
+/*!
+ * User application data buffer size
+ */
+
+#if defined( USE_BAND_915 ) || defined( USE_BAND_915_HYBRID )
+
+#define LORAWAN_APP_DATA_SIZE                       11
+
+#endif
+
+
+extern LoRaMacStatus_t Send(netdev2_lorawan_t *loradev, lw_hdr_t *hdr, uint8_t fPort,
+                             void *fBuffer, uint16_t fBufferSize );
+static uint8_t DevEui[] = LORAWAN_DEVICE_EUI;
+static uint8_t AppEui[] = LORAWAN_APPLICATION_EUI;
+static uint8_t AppKey[] = LORAWAN_APPLICATION_KEY;
+
+static uint8_t NwkSKey[] = LORAWAN_NWKSKEY;
+static uint8_t AppSKey[] = LORAWAN_APPSKEY;
+
+/*!
+ * Device address
+ */
+static uint32_t DevAddr = LORAWAN_DEVICE_ADDRESS;
+
+/*!
+ * Indicates if the MAC layer has already joined a network.
+ */
+static bool IsNetworkJoined = false;
+
+/*!
+ * User application data buffer size
+ */
+#define LORAWAN_APP_DATA_MAX_SIZE                           64
+
+/*!
+ * Indicates if the node is sending confirmed or unconfirmed messages
+ */
+static uint8_t IsTxConfirmed = LORAWAN_CONFIRMED_MSG_ON;
+
+/*!
+ * LoRaWAN compliance tests support data
+ */
+struct ComplianceTest_s
+{
+    bool Running;
+    uint8_t State;
+    bool IsTxConfirmed;
+    uint8_t AppPort;
+    uint8_t AppDataSize;
+    uint8_t *AppDataBuffer;
+    uint16_t DownLinkCounter;
+    bool LinkCheck;
+    uint8_t DemodMargin;
+    uint8_t NbGateways;
+}ComplianceTest;
+
+int lorawan_setup(int argc, char **argv) {
+
+    if (argc < 5) {
+        puts("ERROR: Not enough arguments");
+        return -1;
+    }
+
+    kernel_pid_t pid = atoi(argv[1]);
```

Comment:
Why we should pass netdev's thread's PID into command explicitly? The description of command doesn't implies that.

I think the PID returned from `gnrc_lorawan_init` function called during `auto-init` procedure should be saved somewhere to be accessible from application code.